sockeye/beam_search.py

```diff
@@ -622,21 +624,24 @@ def forward(self,
         # target vocab for this sentence.
         if restrict_lexicon:
             source_words = source[:, :, 0]
-            vocab_slice_ids, _ = _get_vocab_slice_ids(restrict_lexicon, source_words, self.eos_id, beam_size=1, target_prefix=target_prefix)
+            vocab_slice_ids, _ = _get_vocab_slice_ids(restrict_lexicon, source_words, self.eos_id,
+                                                      beam_size=1, target_prefix=target_prefix)
 
         # (0) encode source sentence, returns a list
         model_states, _ = self._inference.encode_and_initialize(source, source_length)
         # TODO: check for disabled predicted output length
 
         # Prefix masks, where scores are infinity for all other vocabulary items except target_prefix ids
-        prefix_masks, prefix_masks_length = (utils.gen_prefix_masking(target_prefix, self.output_vocab_size, self.dtype), target_prefix.size(1)) \
-                                            if target_prefix is not None else (None, None)  # type: ignore
-        if prefix_masks is not None and vocab_slice_ids is not None:
-            prefix_masks = pt.index_select(prefix_masks, -1, vocab_slice_ids)
+        prefix_masks, prefix_masks_length = None, 0
+        if prefix_masks is not None:
+            prefix_masks, prefix_masks_length = utils.gen_prefix_masking(target_prefix, self.output_vocab_size, self.dtype)
+            if vocab_slice_ids is not None:
+                prefix_masks = pt.index_select(prefix_masks, -1, vocab_slice_ids)
         # Prefix factor masks, where scores are also infinity for all other factor items except target_prefix_factor ids
-        target_prefix_factor_masks, target_prefix_factor_length = (utils.gen_prefix_masking(target_prefix_factors, self.output_factor_vocab_size, self.dtype), target_prefix_factors.size(1)) \
-                                                                  if self.num_target_factors > 1 and target_prefix_factors is not None \
-                                                                  else (None, None)  # type: ignore
+        target_prefix_factor_masks, target_prefix_factor_length = None, 0
+        if target_prefix_factors is not None:
```

Comment:
self.num_target_factors > 1 is used for if to handle the case where prefix factors are added (by mistake) but the trained model does not actually has factors. 

Perhaps if we ignore  self.num_target_factors > 1 here, it is worthy to check and assign target_prefix_factors to None if factors is not used in inference.py  in the first place? (somewhere in the function target_prefix_np)

Reply:
Yes, you are right. I somehow thought we check for this already in `make_input_from_dict`. I changed the code to make sure we have the same number of target factor prefixes as there are model-supported target factors. I also removed the warning for an empty target factor because otherwise the user wouldn't be able to supply a prefix for the 3rd factor, but not for the 2nd factor.